src/main/java/frc/robot/subsystems/Vision.java

```diff
@@ -4,46 +4,62 @@
 
 package frc.robot.subsystems;
 
-
-import edu.wpi.first.wpilibj2.command.SubsystemBase;
-import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
 import edu.wpi.first.networktables.NetworkTable;
 import edu.wpi.first.networktables.NetworkTableEntry;
 import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
 
 public class Vision extends SubsystemBase {
   private NetworkTable table;
-  private NetworkTableEntry tx;
-  private NetworkTableEntry ty;
-  private NetworkTableEntry ta;
+  private NetworkTableEntry validTarget;
+  private NetworkTableEntry xOffset;
+  private NetworkTableEntry targetArea; // proxy for distance from target
+  private NetworkTableEntry targetID;
+  private NetworkTableEntry botpose;
 
   public Vision() {
-   table = NetworkTableInstance.getDefault().getTable("limelight-ups");
-   tx = table.getEntry("tx");
-   ty = table.getEntry("ty");
-   ta = table.getEntry("ta");
+    table = NetworkTableInstance.getDefault().getTable("limelight-ups");
+    validTarget = table.getEntry("tv");
+    targetID  = table.getEntry("tid");
+    xOffset = table.getEntry("tx");
+    targetArea = table.getEntry("ta");
+    botpose = table.getEntry("botpose");
   }
 
-  public void getLimelightValues() {
-    //read values periodically
-    double x = tx.getDouble(0.0);
-    double y = ty.getDouble(0.0);
-    double area = ta.getDouble(0.0);
+  public Pose2d getBotpose() {
+    double[] arr = botpose.getDoubleArray(new double[6]);
+    return new Pose2d(arr[0], arr[1], new Rotation2d(arr[3], arr[4]));
+  }
+
+  public int getTargetID() {
+    return (int) targetID.getInteger(0);
+  }
 
-    //post to smart dashboard periodically
-    SmartDashboard.putNumber("LimelightX", x);
-    SmartDashboard.putNumber("LimelightY", y);
-    SmartDashboard.putNumber("LimelightArea", area);
+  public double getTargetArea() {
+    return targetArea.getDouble(0);
   }
 
-  @Override
-  public void periodic() {
-    // This method will be called once per scheduler run
-    getLimelightValues();
+  public Rotation2d getXOffset() {
+    return Rotation2d.fromDegrees(xOffset.getDouble(0));
   }
 
-  @Override
-  public void simulationPeriodic() {
-    // This method will be called once per scheduler run during simulation
+  public boolean hasTarget() {
+    return validTarget.getDouble(0) == 1;
   }
+
+  public void setMode(CamMode mode) {
+    if(mode == CamMode.VISION) {
+      table.getEntry("camMode").setNumber(0);
+    }else {
+      table.getEntry("camMode").setNumber(1);
+    }
+  }
+
+  public static enum CamMode {
+    VISION,
+    DRIVER
+  }
```

Comment:
enum definitions are best near the top of the class traditionally so they are easier to find